### ISSUE TITLE
Revisions aren't deterministic

### ISSUE THREAD

**Issue by marten-de-vries** [pouchdb/pouchdb#2451]
Revision ids are deterministic in CouchDB: http://guide.couchdb.org/draft/conflicts.html#deterministic

Deterministic revisions prevent unnecessary replication conflicts. Them being currently not implemented also causes failures in the CouchDB test suite when run against pouchdb-server. For example (second assertion):
![crash](https://cloud.githubusercontent.com/assets/1750170/3507166/343ed578-067b-11e4-8193-c2d5030d46be.png)

CouchDB code: https://github.com/apache/couchdb/blob/master/src/couchdb/couch_db.erl#L653
PouchDB code (for new docs, at least): https://github.com/pouchdb/pouchdb/blob/master/lib/utils.js#L143

**Comment by nolanlawson** [comment]
@Rybadour It's not really such a big deal in practice.

TLDR: PouchDB replication is slightly less efficient than CouchDB replication, but the behavior will be largely the same.

Let's say two CouchDB users each make the exact same change to the exact same document at the exact same revision (maybe they both fixed a typo in a wiki article or something). CouchDB will give them the same revision hash.

So when it comes time to replicate, database A will ask database B: "Hey, do you have revision `2-x` for document `foo`?" And database B will respond, "Yup, I've already got it." So no conflicts!

PouchDB, on the other hand, will generate a random string for both. So A will ask B, "Hey, do you have this revision?" And B will always say "no." Then they will have to replicate the two revisions, and the document will be in a conflicting state because it has two revision leafs.

(This will occur for both CouchDB<->PouchDB and PouchDB<->PouchDB. CouchDB does not care that we gave the revision the "wrong" hash, i.e. a hash that was not determined by Erlang's string implementation.)

The impact is that the document will be in a conflicting state, and your app will have to resolve a conflict where the two changes are actually the same. But since this is 1) a very unlikely edge case, and 2) something that you can handle yourself in your own conflict-resolution code, it's not really an end-of-the-world scenario for PouchDB.

**Comment by Rybadour** [comment]
Ya I get that it's not a show stopper but it's certainly a point against PouchDB. I was hoping to be able to compare revisions in order to allow our users to respond to conflicts in synchronized data.

**Comment by karlwestin** [comment]
I was asked to add my user story here, i hope it can contribute something.
We're working with very slow network connections, and have about 50 clients using an 'enterprise-style' data entry app. The clients replicate from couchdb to pouchdb, resolve conflicts that might have occured, and then they attempt to replicate back. This may or may not work depending on the network connection they have at the moment. This conflict resolution code is deterministic, and creates the same result, so we'll basically have 2 of the same documents, but with different revision numbers. Now replicating those back will create another conflict, which resolves deterministically again, and as long as clients are syncing at the same time, so on. This is not a huge issue, we're pretty much just running the risk of creating some new revisions until one client can sync alone. I'm just noting that some deterministic revisions (even though it wouldnt be exactly the same as couchdbs) would be helpful here.

**Comment by nolanlawson** [comment]
Thanks for the input! :smiley: 

I think we would accept a pull request for an option to do deterministic revision handling (inconsistent with CouchDB, but consistent with PouchDB), but it would probably have to be false by default because that would entail doing an expensive checksum for every revision. 

If anyone is interested in contributing, the code that would need to be changed is [here](https://github.com/pouchdb/pouchdb/blob/9c093980d4bc03b19d9dfdb338f656849ee01a6b/lib/deps/parse-doc.js#L109). We already have an md5 library bundled with PouchDB, which is in `utils.md5`. And you would have to pass in an option to `parseDoc`, which could maybe be set globally on the `PouchDB` object or, ideally, on the `db` instance itself.

**Comment by npow** [comment]
Just ran into this today. @daleharvey can you link me to those previous patches to match couchdb's revision generation?

**Comment by daleharvey** [comment]
Closing as a dupe - https://github.com/pouchdb/pouchdb/issues/4642
